### PR TITLE
Complete multisig support

### LINKED ISSUE #248: coldcard multisig testing coverage
we need it

@justinmoon you said there's a bug currently with native segwit outputs? may want to wait until that is fixed.

### PR DESCRIPTION
**Pull request by achow101**
I've tested multisig with all of the devices supported by HWI. This PR fixes the implementations in devices where multisig did not work as well as adds tests for multisig as applicable.

Only two devices had issues with multisig: Trezor T and Coldcard. The Trezor T has some issues because of it's inability to have mixed segwit and non-segwit inputs in a transaction. This poses an issue for us because we skip inputs by declaring the input as being a segwit input and then discarding the signature for it. In this PR, this issue is now documented as a caveat. However this did cause some problems with multisigs since we have to do multiple passes of the transaction, and later rounds would originally mark fully signed inputs as not ours, which, when having legacy multisigs, runs into the mixed segwit issue. This is resolved by not marking those as segwit inputs but still ignoring them.

Additionally, there is now a supports external inputs whitelist in the test to avoid issues with that.

To use multisig with the Coldcard requires enrolling the multisig wallet beforehand. This requirement is now documented. For our tests, we enroll the multisig wallet at startup using the method provided by the simulator's command line arguments. We apply a patch to the simulator to use the multisig wallets that we use in the tests instead of the default provided by the simulator. Additionally, like the Trezor, the Coldcard needs to do multiple passes of signing the transaction if there is a multisig in which the Coldcard is a participant multiple times. So the `sign_tx()` function is updated to handle the multiple passes. Lastly, the Coldcard uses and enforces BIP 67. So we need to use the `sortedmulti` descriptor to handle that.

Both the Coldcard and Trezor T are now enabled for multisig tests.

Closes #248

### CODE REVIEW

File: hwilib/devices/coldcard.py [hunk 1]
```diff
@@ -68,53 +69,72 @@ def _get_fingerprint_hex(self):
...
-        left = sz
-        chk = sha256()
-        for pos in range(0, sz, MAX_BLK_LEN):
-            here = fd.read(min(MAX_BLK_LEN, left))
-            if not here:
+        # Get this devices master key fingerprint
+        xpub = self.device.send_recv(CCProtocolPacker.get_xpub('m/0\''), timeout=None)
+        master_fp = get_xpub_fingerprint(xpub)
+
+        # For multisigs, we may need to do multiple passes if we appear in an input multiple times
```
Comment (instagibbs) [inline hwilib/devices/coldcard.py]: oof, what does this look like for the user? multiple independent signing passes?

Comment (instagibbs) [reply hwilib/devices/coldcard.py]: also, this needs a test

File: hwilib/devices/trezor.py [hunk 2]
```diff
@@ -220,7 +224,7 @@ def ignore_input():
                 if our_keys > passes:
                     passes = our_keys
 
-                if not found:
+                if not found and not found_in_sigs:
```
Comment (instagibbs) [inline hwilib/devices/trezor.py]: can you add some more comments/explanation in the code for this logic? I don't think this is obvious.

File: docs/trezor.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ Due to the limitations of the Trezor, some transactions cannot be signed by a Tr
 - Multisig inputs are limited to at most n-of-15 multisigs. This is a firmware limitation.
 * Transactions with arbitrary input scripts (scriptPubKey, redeemScript, or witnessScript) and arbitrary output scripts cannot be signed. This is a firmware limitation.
 * Send-to-self transactions will result in no prompt for outputs as all outputs will be detected as change.
+- For **Trezor T**, a transaction cannot contain both segwit and non-segwit inputs
```
Comment (instagibbs) [inline docs/trezor.md]: what happens if it does?

File: test/test_device.py [hunk 4]
```diff
@@ -361,9 +361,9 @@ def _test_signtx(self, input_type, multisig):
                    pkh_info['desc'][4:-10]]
 
         # Get the descriptors with their checksums
-        sh_multi_desc = self.wrpc.getdescriptorinfo('sh(multi(2,' + pubkeys[0] + ',' + pubkeys[1] + ',' + pubkeys[2] + '))')['descriptor']
-        sh_wsh_multi_desc = self.wrpc.getdescriptorinfo('sh(wsh(multi(2,' + pubkeys[0] + ',' + pubkeys[1] + ',' + pubkeys[2] + ')))')['descriptor']
-        wsh_multi_desc = self.wrpc.getdescriptorinfo('wsh(multi(2,' + pubkeys[2] + ',' + pubkeys[1] + ',' + pubkeys[0] + '))')['descriptor']
+        sh_multi_desc = self.wrpc.getdescriptorinfo('sh(sortedmulti(2,' + pubkeys[0] + ',' + pubkeys[1] + ',' + pubkeys[2] + '))')['descriptor']
```
Comment (instagibbs) [inline test/test_device.py]: note that this functionality is only in Core master branch for now, so support is a little shaky until 0.20. This should probably go in notes somewhere that this is required for use with Core.

File: hwilib/devices/coldcard.py (hunk 1 above)
Comment (achow101) [reply hwilib/devices/coldcard.py]: It is multiple independent signings. There already are tests for this, our multisig tests use 3 keys from the same device. This was needed to pass our test suite.

File: docs/trezor.md (hunk 3 above)
Comment (achow101) [reply docs/trezor.md]: You get an error that says `Some signatures are missing`

File: hwilib/devices/coldcard.py (hunk 1 above)
Comment (instagibbs) [reply hwilib/devices/coldcard.py]: ah i see github was confusing me with commit ordering, looks good

File: hwilib/devices/trezor.py (hunk 2 above)
Comment (achow101) [reply hwilib/devices/trezor.py]: Added some comments.

File: test/test_device.py (hunk 4 above)
Comment (achow101) [reply test/test_device.py]: Added a note in the coldcard signtx caveats.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin-core/HWI/pull/289/commits/fc0b01f70a03c63009d140d372a6dfef74b389be

**Comment by achow101** [comment]
Is anyone able to replicate the travis failures? I can't get those tests to fail locally.

**Comment by achow101** [comment]
Finally figured out what was causing the travis error.

**Comment by instagibbs** [comment]
re-utACK https://github.com/bitcoin-core/HWI/pull/289/commits/c9107b03c9da1b2a391c9ac674400247071c6caf

**Pull request merged**